### ISSUE TITLE
`weechat.command('', '/msg #lol text')` sent twice when triggered by hook `irc_out_PRIVMSG`

### ISSUE THREAD

**Issue by al-caveman** [weechat/weechat#1131]
WeeChat 2.0.1 [compiled on Jan  3 2018 08:05:45], obtined from Arch Linux's `community` repository.

Do this:

 1. Copy contents of this page https://github.com/Al-Caveman/weechatbots/tree/master/holybooks into your local `~/.weechat/python`.
 2. Edit the file `holybooks.py` to ensure that `CHANNELS_ACTIVE` contains the right `network/#channel` where you wish to test this. In my case, `freenode` is the name of the Freenode server, and I added a few channels. So you need to adjust those to match whatever is your case.
 3. In `weechat`, run `/script load holybooks.py`.
 4. Now, join a channel in `CHANNELS_ACTIVE`.
 5. Then, using another IRC client, also join the same channel over another connection that is not through the same `weechat` instance. Perhaps, using the web Freenode feature?
 6. Using the other IRC client (e.g. maybe web Freenode) type `!genesis 1:1` and you will find the script work normally as intended (prints the 1st verse of the 1st chapter in the book of genesis.
 7. Now, using your `weechat` that is running the `holybooks.py` script, enter that channel, and type the same command `!genesis 1:1`. What you will see on your end is similar: the 1st verse of the 1st chapter of the book of genesis.
 8. Here is the problem: if you look at the other IRC client (e.g. web Freenode), you will see that the script seems to have sent two copies of the Genesis verse *only* when you type `!genesis 1:1` locally in `weechat` where the script is running.

 Question is: why are there two copies in step 8?

In other words, it seems that the verse is sent once (good) when triggered by the hook `wc.hook_signal('*,irc_in2_PRIVMSG', 'callback', '')`, but is sent twice (bad) when triggered by the hook `wc.hook_signal('*,irc_out_PRIVMSG', 'callback', '')`.

**Comment by Al-Caveman** [comment]
WeeChat 2.0.1 [compiled on Jan  3 2018 08:05:45]

**Comment by flashcode** [comment]
Hi,
When you do `wc.command('', '/msg -server ...")`, this triggers again your callback on signal "irc_out_PRIVMSG", so it may call again the callback while you are still inside.
WeeChat has some mechanism to protect against infinite loop on callbacks call for the same signal, but you should add a mechanism in your script to protect against that (ie if the callback is called again while another one is running, you should do nothing).

**Comment by sim642** [comment]
As a side note: in IRC callbacks you should **always** use [`irc_message_parse`](https://weechat.org/files/doc/devel/weechat_plugin_api.en.html#_info_get_hashtable) to parse the message, not your own splitting and regexes.

**Comment by Al-Caveman** [comment]
@flashcode but for the subsequent `irc_out_PRIVMSG` that's triggered with `wc.command('', '/msg -server ...")`, the content of the message is different, and hence the function callback won't send anything.

In other words, the function callback sends messages only when the read text follows the correct syntax `!BOOK X:Y`. To be more certain, I put a `wc.prnt('', 'SENDING')` in the sending function, which causes it to pring `SENDING` everytime the callback results in sending something. The outcome of my tests so far indicate that `SENDING` is printed only once.

So I still don't understand why a 2n callback should result in sending anything.

**Comment by flashcode** [comment]
So please try to add debug prints in the script (with `weechat.prnt("", "some debug")`), to find where something is wrong (either in script or WeeChat).

It's better if you can say exactly the bug you think there is in WeeChat, it's not easy for me to debug user scripts (not enough time on my side).

**Comment by Al-Caveman** [comment]
Line 102 in https://github.com/Al-Caveman/weechatbots/blob/master/holybooks/holybooks.py already has `wc.prnt('', 'SENDING')`.

I only have a few guesses and I hope they help:

  1. server raw shows a single message. So I guess this means that a single message is sent.
  2. But , other servers are getting 2 messages? W.a.t.

I think of two options: 
  1. Either this is a very fundamental bug in weechat that not even server raw can detect.
  2. Or that it is something remote servers are doing.

To be sure which is which, I will connect over plain text and test via wireshark. Will be back when this happens.

**Comment by Al-Caveman** [comment]
OK, after testing with wireshark:

  1. The 1st text that I send, namely `! genesis 1:1` is being used as expected to send the actual Biblical verse. So, the biblical verse is sent.
  2. But, then, the original text `! genesis 1:1` is not sent, and -instead- is replaced by the Biblical verse.

This is why I am sending two Biblical verses (because the command `!genesis 1:1` is also replaced by a verse itself.

I confirm that with Wireshark.

**Comment by flashcode** [comment]
Hi,
Please give me the exact steps to reproduce the issue, or a script showing the problem.
Thanks.

**Comment by Al-Caveman** [comment]
Hi - steps:

1. Clone this git repository https://github.com/Al-Caveman/weechatbots/blob/master/holybooks/
2. Augment variable `CHANNELS_ACTIVE` in `holybooks.py` by adding the network/channel where the script shall be active.
3. Load `holybooks.py` as a python script.
4. Go to that channel, and type `!quran 1:1`.
5. You should see the verse printed twice.

**Notes:**

* I'm only replying here to answer your request, under the assumption that maybe you are interested in chasing this, in case it helps fixing an actual bug.
* This is not a high priority at all for me, and I no longer use the script.  Please feel free to postpone this if you feel so, or even closing.

**Comment by flashcode** [comment]
Closing old issue, if you still have the problem with the latest version, please add a comment.
